File: Assets/Bundles/Server/Resources/TS/Services/Match/BedService.lua
-- Compiled with unity-ts v2.1.0-75
local Reflect = require("Shared/rbxts_include/node_modules/@easy-games/flamework-core/out/init").Reflect
local Service = require("Shared/rbxts_include/node_modules/@easy-games/flamework-core/out/init").Service
local ObjectUtils = require("Shared/rbxts_include/node_modules/@easy-games/unity-object-utils/init")
local ServerSignals = require("Server/TS/ServerSignals").ServerSignals
local GetItemMeta = require("Shared/TS/Item/ItemDefinitions").GetItemMeta
local ItemType = require("Shared/TS/Item/ItemType").ItemType
local MathUtil = require("Shared/TS/Util/MathUtil").MathUtil
local Task = require("Shared/TS/Util/Task").Task
local VoxelDataAPI = require("Shared/TS/VoxelWorld/VoxelData/VoxelDataAPI").VoxelDataAPI
local WorldAPI = require("Shared/TS/VoxelWorld/WorldAPI").WorldAPI
-- * Bed block id.
local _result = GetItemMeta(ItemType.BED).block
if _result ~= nil then
	_result = _result.blockId
end
local _condition = _result
if _condition == nil then
	_condition = -1
end
local BED_BLOCK_ID = _condition
local BedService
do
	BedService = setmetatable({}, {
		__tostring = function()
			return "BedService"
		end,
	})
	BedService.__index = BedService
	function BedService.new(...)
		local self = setmetatable({}, BedService)
		return self:constructor(...) or self
	end
	function BedService:constructor(mapService, matchService)
		self.mapService = mapService
		self.matchService = matchService
		self.teamToBed = {}
	end
	function BedService:OnStart()
		-- Listen for bed destroyed.
		ServerSignals.BeforeBlockDestroyed:Connect(function(event)
			if event.blockId == BED_BLOCK_ID then
				local teamId = VoxelDataAPI:GetVoxelData(event.blockPos, "teamId")
				if not (teamId ~= "" and teamId) then
					return nil
				end
				ServerSignals.BedDestroyed:Fire({
					bedTeamId = teamId,
				})
			end
		end)
		-- Spawn beds after map load and match start.
		Task:Spawn(function()
			self.loadedMap = self.mapService:WaitForMapLoaded()
			ServerSignals.MatchStart:connect(function()
				return self:SpawnBeds()
			end)
		end)
	end
	function BedService:SpawnBeds()
		-- Spawn beds.
		local beds = self.loadedMap:GetAllBeds()
		local _exp = ObjectUtils.keys(beds)
		local _arg0 = function(teamId)
			local bed = beds[teamId]
			local bedPos = MathUtil:FloorVec(Vector3.new(bed.Position.x, bed.Position.y, bed.Position.z))
			local bedState = {
				teamId = teamId,
				position = bedPos,
				destroyed = false,
			}
			local _teamToBed = self.teamToBed
			local _teamId = teamId
			_teamToBed[_teamId] = bedState
			WorldAPI:GetMainWorld():PlaceBlock(bedPos, ItemType.BED)
			-- TEMPORARY. Fix `VoxelDataAPI` race condition.
			Task:Delay(1, function()
				VoxelDataAPI:SetVoxelData(bedPos, "teamId", teamId)
			end)
		end
		for _k, _v in _exp do
			_arg0(_v, _k - 1, _exp)
		end
	end
	function BedService:IsBedDestroyed(teamId)
		local bedState = self:GetBedStateForTeamId(teamId)
		if not bedState then
			return true
		end
		return bedState.destroyed
	end
	function BedService:GetBedStateForTeamId(teamId)
		local _teamToBed = self.teamToBed
		local _teamId = teamId
		return _teamToBed[_teamId]
	end
end
-- (Flamework) BedService metadata
Reflect.defineMetadata(BedService, "identifier", "Bundles/Server/Services/Match/BedService@BedService")
Reflect.defineMetadata(BedService, "flamework:parameters", { "Bundles/Server/Services/Match/Map/MapService@MapService", "Bundles/Server/Services/Match/MatchService@MatchService" })
Reflect.defineMetadata(BedService, "flamework:implements", { "$:flamework@OnStart" })
Reflect.decorate(BedService, "$:flamework@Service", Service, { {} })
return {
	BedService = BedService,
}
-- ----------------------------------
-- ----------------------------------

File: Assets/Typescript~/src/Server/Services/Match/BedService.ts
import { CoreServerSignals } from "@Easy/Core/Server/CoreServerSignals";
import { TeamService } from "@Easy/Core/Server/Services/Team/TeamService";
import { Entity } from "@Easy/Core/Shared/Entity/Entity";
import { Game } from "@Easy/Core/Shared/Game";
import { ItemType } from "@Easy/Core/Shared/Item/ItemType";
import { ItemUtil } from "@Easy/Core/Shared/Item/ItemUtil";
import { Team } from "@Easy/Core/Shared/Team/Team";
import { ColorUtil } from "@Easy/Core/Shared/Util/ColorUtil";
import { MathUtil } from "@Easy/Core/Shared/Util/MathUtil";
import { Theme } from "@Easy/Core/Shared/Util/Theme";
import { BlockDataAPI, CoreBlockMetaKeys } from "@Easy/Core/Shared/VoxelWorld/BlockData/BlockDataAPI";
import { WorldAPI } from "@Easy/Core/Shared/VoxelWorld/WorldAPI";
import { OnStart, Service } from "@easy-games/flamework-core";
import { ServerSignals } from "Server/ServerSignals";
import { BedState } from "Shared/Bed/BedMeta";
import { MapService } from "./Map/MapService";
import { MatchService } from "./MatchService";

/** Bed block id. */
const BED_BLOCK_ID = ItemUtil.GetItemMeta(ItemType.BED).block?.blockId ?? -1;

@Service({})
export class BedService implements OnStart {
	/** Team id to bed map. */
	private teamToBed = new Map<string, BedState>();

	constructor(
		private readonly mapService: MapService,
		private readonly matchService: MatchService,
		private readonly teamService: TeamService,
	) {}

	OnStart(): void {
		/* Listen for bed destroyed. */
		CoreServerSignals.BeforeBlockDestroyed.Connect((event) => {
			if (event.blockId === BED_BLOCK_ID) {
				this.TryDestroyBed(event.blockPos, event.entity);
			}
		});

		ServerSignals.MatchStart.Connect(() => {
			this.SpawnBeds();
		});
	}

	private TryDestroyBed(blockPos: Vector3, entity?: Entity): boolean {
		const teamId = BlockDataAPI.GetBlockData<string>(blockPos, "teamId");
		if (!teamId) return false;
		const team = this.teamService.GetTeamById(teamId);
		if (team) {
			const breakerTeam = entity?.player?.GetTeam();
			if (entity && breakerTeam) {
				Game.BroadcastMessage(
					ColorUtil.ColoredText(breakerTeam.color, entity.GetDisplayName()) +
						ColorUtil.ColoredText(Theme.Aqua, " destroyed ") +
						ColorUtil.ColoredText(team.color, `${team.name} team's bed`) +
						ColorUtil.ColoredText(Theme.Aqua, "!"),
				);
			} else {
				Game.BroadcastMessage(
					ColorUtil.ColoredText(team.color, `${team.name}'s bed`) +
						ColorUtil.ColoredText(Theme.Aqua, " has been destroyed!"),
				);
			}
			const bedState = this.GetBedStateForTeam(team);
			if (bedState) {
				bedState.destroyed = true;
			}
			ServerSignals.BedDestroyed.Fire({ team });
			return true;
		}
		return false;
	}

	/** Spawn beds for each team. */
	private SpawnBeds(): void {
		const loadedMap = this.mapService.GetLoadedMap();
		if (!loadedMap) return;

		// Spawn beds.
		for (let team of this.teamService.GetTeams()) {
			const bedPosition = loadedMap.GetWorldPosition(team.id + "_bed");
			const bedPos = MathUtil.FloorVec(
				new Vector3(bedPosition.Position.x, bedPosition.Position.y, bedPosition.Position.z),
			);
			const bedState: BedState = {
				teamId: team.id,
				position: bedPos,
				/* _Always_ starts as not destroyed. */
				destroyed: false,
			};
			this.teamToBed.set(team.id, bedState);
			const itemMeta = ItemUtil.GetItemMeta(ItemType.BED);
			WorldAPI.GetMainWorld()!.PlaceBlockByItemType(bedPos, ItemType.BED, {
				blockData: {
					teamId: team.id,
					health: itemMeta.block!.health!, // this is a hack.
					[CoreBlockMetaKeys.CAN_BREAK]: true,
				},
			});
		}
	}

	/**
	 * Checks whether or not a team's bed is destroyed.
	 * @param teamId A team id.
	 * @returns Whether or not a specific team's bed is destroyed.
	 */
	public IsBedDestroyed(team: Team): boolean {
		const bedState = this.GetBedStateForTeam(team);
		if (!bedState) return true;
		return bedState.destroyed;
	}

	/**
	 * Fetch bed state with a team id.
	 * @param teamId A team id.
	 * @returns Bed state.
	 */
	public GetBedStateForTeam(team: Team): BedState | undefined {
		return this.teamToBed.get(team.id);
	}
}
